README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# bpnreg

[![CRAN_Status_Badge](http://www.r-pkg.org/badges/version/bpnreg)](https://cran.r-project.org/package=bpnreg)

The goal of bpnreg is to fit Bayesian projected normal regression models
for circular data.

## Installation

The R-package bpnreg can be installed from CRAN as follows:

``` r
install.packages("bpnreg")
```

You can install a beta-version of bpnreg from github with:

``` r
# install.packages("devtools")
devtools::install_github("joliencremers/bpnreg")
```

## Citation

To cite the package ‘bpnreg’ in publications use:

Jolien Cremers (2021). bpnreg: Bayesian Projected Normal Regression
Models for Circular Data. R package version 2.0.2.
<https://CRAN.R-project.org/package=bpnreg>

## Example

This is a basic example which shows you how to run a Bayesian projected
normal regression model:

``` r
library(bpnreg)
bpnr(Phaserad ~ Cond + AvAmp, Motor, its = 100)
#> Iteration:1
#> Iteration:2
#> Iteration:3
#> Iteration:4
#> Iteration:5
#> Iteration:6
#> Iteration:7
#> Iteration:8
#> Iteration:9
#> Iteration:10
#> Iteration:11
#> Iteration:12
#> Iteration:13
#> Iteration:14
#> Iteration:15
#> Iteration:16
#> Iteration:17
#> Iteration:18
#> Iteration:19
#> Iteration:20
#> Iteration:21
#> Iteration:22
#> Iteration:23
#> Iteration:24
#> Iteration:25
#> Iteration:26
#> Iteration:27
#> Iteration:28
#> Iteration:29
#> Iteration:30
#> Iteration:31
#> Iteration:32
#> Iteration:33
#> Iteration:34
#> Iteration:35
#> Iteration:36
#> Iteration:37
#> Iteration:38
#> Iteration:39
#> Iteration:40
#> Iteration:41
#> Iteration:42
#> Iteration:43
#> Iteration:44
#> Iteration:45
#> Iteration:46
#> Iteration:47
#> Iteration:48
#> Iteration:49
#> Iteration:50
#> Iteration:51
#> Iteration:52
#> Iteration:53
#> Iteration:54
#> Iteration:55
#> Iteration:56
#> Iteration:57
#> Iteration:58
#> Iteration:59
#> Iteration:60
#> Iteration:61
#> Iteration:62
#> Iteration:63
#> Iteration:64
#> Iteration:65
#> Iteration:66
#> Iteration:67
#> Iteration:68
#> Iteration:69
#> Iteration:70
#> Iteration:71
#> Iteration:72
#> Iteration:73
#> Iteration:74
#> Iteration:75
#> Iteration:76
#> Iteration:77
#> Iteration:78
#> Iteration:79
#> Iteration:80
#> Iteration:81
#> Iteration:82
#> Iteration:83
#> Iteration:84
#> Iteration:85
#> Iteration:86
#> Iteration:87
#> Iteration:88
#> Iteration:89
#> Iteration:90
#> Iteration:91
#> Iteration:92
#> Iteration:93
#> Iteration:94
#> Iteration:95
#> Iteration:96
#> Iteration:97
#> Iteration:98
#> Iteration:99
#> Iteration:100
#> Projected Normal Regression 
#> 
#> Model 
#> 
#> Call: 
#> bpnr(pred.I = Phaserad ~ Cond + AvAmp, data = Motor, its = 100)
#> 
#> MCMC: 
#> iterations = 100
#> burn-in = 1
#> lag = 1
#> 
#> Model Fit: 
#>         Statistic Parameters
#> lppd    -57.22945   8.000000
#> DIC     127.66465   6.768024
#> DIC.alt 124.17298   5.022188
#> WAIC1   127.33436   6.437733
#> WAIC2   128.65389   7.097498
#> 
#> 
#> Linear Coefficients 
#> 
#> Component I: 
#>                      mean        mode         sd      LB HPD     UB HPD
#> (Intercept)   1.319611894  1.39128370 0.45635201  0.33485506 2.03794238
#> Condsemi.imp -0.522451171 -0.47667290 0.57057933 -1.55833243 0.50939839
#> Condimp      -0.650053029 -0.99688228 0.64741848 -2.00362696 0.53197461
#> AvAmp        -0.009320081 -0.01808984 0.01296947 -0.03096035 0.01524266
#> 
#> Component II: 
#>                     mean         mode         sd     LB HPD       UB HPD
#> (Intercept)   1.37081341  1.057909990 0.43448499  0.5256653  2.265534446
#> Condsemi.imp -1.13529041 -1.508829276 0.60583443 -2.2586284  0.029840305
#> Condimp      -0.93550260 -1.263941265 0.62075876 -2.3158274 -0.009041090
#> AvAmp        -0.01016616 -0.003931414 0.01062028 -0.0285245  0.008526117
#> 
#> 
#> Circular Coefficients 
#> 
#> Continuous variables: 
#>    mean ax    mode ax      sd ax      LB ax      UB ax 
#>  116.31973   76.25854  562.60196 -154.19115  219.74298 
#> 
#>    mean ac    mode ac      sd ac      LB ac      UB ac 
#>  1.0746179  2.2543777  1.1994513 -0.8224601  2.4169745 
#> 
#>      mean bc      mode bc        sd bc        LB bc        UB bc 
#> -0.034814814 -0.006854753  0.499046459 -0.767238134  0.666230333 
#> 
#>       mean AS       mode AS         sd AS         LB AS         UB AS 
#>  4.875002e-04  6.466495e-05  5.442953e-03 -1.160784e-02  2.842468e-03 
#> 
#>     mean SAM     mode SAM       sd SAM       LB SAM       UB SAM 
#> 1.437848e-03 1.305745e-04 1.940441e-02 3.180594e-08 3.466995e-03 
#> 
#>   mean SSDO   mode SSDO     sd SSDO     LB SSSO     UB SSDO 
#> -0.05101017  1.88339563  1.99577431 -2.77725635  2.64369230 
#> 
#> Categorical variables: 
#> 
#> Means: 
#>                           mean       mode        sd         LB        UB
#> (Intercept)          0.8119255  0.8675846 0.1957991  0.4326112 1.2082844
#> Condsemi.imp         0.2962062  0.3373583 0.3399843 -0.4996824 0.8360214
#> Condimp              0.5851581  0.4454521 0.4819606 -0.4032866 1.4047517
#> Condsemi.impCondimp -1.3273542 -2.0443304 1.1135480 -2.8870086 1.4407720
#> 
#> Differences: 
#>                          mean      mode        sd         LB       UB
#> Condsemi.imp        0.5152442 0.4826193 0.4033441 -0.2197928 1.286760
#> Condimp             0.2261741 0.3480214 0.5484078 -0.8033373 1.395936
#> Condsemi.impCondimp 2.2043432 2.8593855 1.0362019 -0.4035095 3.855837
```


NEWS.md
# bpnreb 2.0.3.9000

This is a development version

# bpnreg 2.0.3

* Small documentation fixes for the breaking change:

Using @docType package no longer automatically adds a -package alias. Instead document _PACKAGE to get all the defaults for package documentation.

* Documentation fix for unused arguments in b_samp and betaBlock

* Update JSS hyperlinks

# bpnreg 2.0.2

## Bug fixes
* Made sure that formulae for SAM, AS and bc correspond to those in Cremers, Mulder \& Klugkist (2018). *Circular interpretation of regression coefficients.* In previous versions of the package there was an error in the computation of SAM, AS and bc causing erroneous results.

* Corrected c++ code of mcmc sampler for intercept-only circular mixed effects model to solve error warning in v.2.0.0 and v.2.0.1.

* Corrected incorrect computation of model fit estimates for circular mixed effects models in v.2.0.0 and v.2.0.1.

* Made sure that the chosen lag is displayed correctly in the output.

# bpnreg 2.0.1

## Bug fixes
* Removed MASS from imports to fix note in the 2.0.0 version

* Changes in c++ code to solve errors from CRAN valgrind checks in the 2.0.0 version.

* Changed arma::inv_sympd to arma::inv in c++ code to solve errors on macOS in the 2.0.0 version.

# bpnreg 2.0.0
## Major changes
* Changed the names of the mcmc output objects. `B1`, `B2`, `Beta.I`, `Beta.II`, `B.I`, `B.II`, `VCov.I` and `VCov.II` are now named `beta1`, `beta2`, `beta1`, `beta2`, `b1`, `b2`, `omega1` and `omega2` respectively.

* Functions `predict()` and `residual()` were removed.

* MCMC sampler underlying the `bpnme` function is rewritten in `c++`. This should significantly increase computational speed.

* Increased memory efficiency by not storing intermediate objects in memory.

* A vignette containing answers to frequently asked questions about usage of the package is included.

## Bug fixes
* Adapted documentation `bpnme` function such that it clearly states only one grouping factor can be used.

* Made sure that the circular random intercept variance (and circular random slope variance of categorical predictors) is computed as 1 - mean resultant length. Also updated documentation accordingly.

* Results for `bc`, `AS` and `SAM`  are now also included in degrees if `units = "degrees"` in `circ_coef` function

* An error warning in case of any missing values was included.

* The `WAIC` criterion in the model fit estimates obtained from the `fit()` function is more correctly referred to as `WAIC1`.

* Format of `cRS` object stored in a `bpnme` fit object was changed from `matrix` to `data.frame` to accommodate for data of more than two types, e.g. `character` and `numeric`.

# bpnreg 1.0.3
* Deleted arguments from function documentation that are not in `\usage`.

* Updated references in the packages description.

# bpnreg 1.0.2
* This version contains an update of code containing the class function to make sure the package works with R version 4.0.0

* The check for the range of the circular outcome was corrected.

# bpnreg 1.0.1

* The function bpnme can now also handle the standard R data.frame instead of only Data Frame Tbl (from dplyr, tbl_df(data))

* The function coef.bpnr now correctly returns radians instead of degrees when option units == "radians" is chosen.

* The documentation now states that the circular outcome needs to be measured in radians.

* The references in the documentation have been updated.

* The functions mmr and mmme now give out a warning message when the circular outcome is out of range.

* An error in the assignment of column names in the random effect model matrix (in mmme function) was fixed.

* An error in the check for whether random effects are numeric (in bpnme function) was fixed.

* Several checks and error warnings for wrong specifications of nesting structure were added.

* Internal functions were taken out of the package documentation on CRAN.

* The priors for the (fixed effect) regression coefficients in the regression and mixed-effects model are now both N(0, 10000), We also note this in the documentation of bpnr() and bpnme().

* We note the fact that the mixed-effects model is only developed for models with one nesting variable in the documentation.

# bpnreg 1.0.0
* This is the first version of the package.


R/regression.R
#' Fit a Bayesian circular regression model
#'
#' This function fits a Bayesian circular regression model based on the
#' projected normal distribution.
#'
#' @param pred.I model equation for effects of component 1.
#' @param data the dataframe used for analysis.
#' @param pred.II model equation for effects of component 2.
#' @param its output iterations of the MCMC sampler.
#' @param burn number of burn-in iterations.
#' @param n.lag amount of lag for the iterations and burn-in.
#' @param seed user-specified random seed.
#'
#' @details Because the model is based on the projected normal distribution, a
#'   model equation has to be given for two components. By default the equation
#'   of the second component \code{pred.II} is set to be equal to that of the
#'   first component. Note that the circular outcome needs to be measured in
#'   radians on a scale from 0 to 2\eqn{\pi}. For more information about the
#'   projected normal distribution see Presnell, Morrisson & Littell (1998).
#'
#'   A tutorial on how to use this function can be found in Cremers & Klugkist
#'   (2018). More details on the sampling algorithm and interpretation of the
#'   coefficients from the model can be found in Cremers, Mulder & Klugkist
#'   (2018) and Cremers, Mainhard & Klugkist (2018). The uninformative priors
#'   for the regression coefficients of the two components are set to N(0, 10000).
#'
#' @return A \code{bpnr} object, which can be further analyzed using the
#'   associated functions \code{\link{traceplot.bpnr}}, \code{\link{BFc.bpnr}},
#'   \code{\link{coef_lin.bpnr}}, \code{\link{coef_circ.bpnr}},
#'   \code{\link{fit.bpnr}} and \code{\link{print.bpnr}}.
#'
#'   A \code{bpnr} object contains the following elements (some elements are not
#'   returned if not applicable)
#'
#'   \describe{ \item{\code{beta1}}{A matrix of posterior samples for the
#'   coefficients \code{beta1} of the first component.} \item{\code{beta2}}{A matrix
#'   of posterior samples for the coefficients \code{beta2} for the second
#'   component.} \item{\code{Likelihood}}{A matrix containing the posterior
#'   density values for all individuals in the dataset for all iterations. The
#'   rowsums of this matrix are the likelihood values for all iterations}
#'   \item{\code{its}}{Number of output iterations.} \item{\code{n.lag}}{One in
#'   \code{n.lag} iterations will be saved as output iteration. Set lag to 1 to
#'   save all iterations (default).} \item{\code{burn-in}}{Burn-in time for the
#'   MCMC sampler.} \item{\code{p1}}{Number of parameters predicting the first
#'   component.} \item{\code{p2}}{Number of parameters predicting the second
#'   component.} \item{\code{theta}}{The circular outcome vector measured in radians.}
#'   \item{\code{a.x}}{A matrix of posterior samples for \code{a.x} which
#'   describes the location of the inflection point of the regression curve on
#'   the axis of the predictor.} \item{\code{a.c}}{A matrix of posterior samples
#'   for \code{a.c} which describes the location of the inflection point of the
#'   regression curve on the axis of the circular outcome.} \item{\code{b.c}}{A
#'   matrix of posterior samples for \code{b.c} which describes the slope of the
#'   tangent line at the inflection point.} \item{\code{SAM}}{A matrix of
#'   posterior samples for the circular regression slopes at the mean.}
#'   \item{\code{AS}}{A matrix of posterior samples for the average slopes of
#'   the circular regression.} \item{\code{SSDO}}{A matrix of posterior samples
#'   for the signed shortest distance to the origin.} \item{\code{circ.diff}}{A
#'   matrix of posterior samples for the circular difference between levels of
#'   categorical variables and the intercept.} \item{\code{Call}}{The matched
#'   call.} \item{\code{lin.coef.I}}{The mean, mode, standard deviation and 95 %
#'   confidence interval of the highest posterior density of the linear
#'   coefficients for \code{beta1}.} \item{\code{lin.coef.II}}{The mean, mode,
#'   standard deviation and 95 % confidence interval of the highest posterior
#'   density of the linear coefficients for \code{beta2}.}
#'   \item{\code{circ.coef}}{The mean, mode, standard deviation and 95 %
#'   confidence interval of the highest posterior density for the \code{a.x},
#'   \code{a.c}, \code{b.c}, \code{AS}, \code{SAM} and \code{SSDO} of the
#'   circular coefficients.} \item{\code{circ.coef.cat}}{The mean, mode,
#'   standard deviation and 95 % confidence interval of the highest posterior
#'   density the circular difference between levels of categorical variables and
#'   the intercept.} \item{\code{circ.coef.means}}{The mean, mode, standard
#'   deviation and 95 % confidence interval of the highest posterior density of
#'   circular means of the categorical variables.} \item{\code{model.fit}}{A
#'   list of information criteria for assessment of model fit.}
#'   \item{\code{mm}}{A list of information, model matrices, sample size, etc.
#'   on the specified model.} }
#'
#'
#' @source Cremers, J., Mulder, K.T. & Klugkist, I. (2018). Circular
#'   interpretation of regression coefficients. British Journal of Mathematical
#'   and Statistical Psychology, 71(1), 75-95.
#'
#' @source Cremers, J., Mainhard, M.T. & Klugkist, I. (2018). Assessing a
#'   Bayesian Embedding Approach to Circular Regression Models. Methodology, 14,
#'   69-81.
#'
#' @source Cremers, J. & Klugkist, I. (2018). One direction? A tutorial for
#'   circular data with examples in cognitive psychology. Frontiers in
#'   Psychology: Cognitive Science.
#'
#' @source Presnell, B., Morrison, S.P. & Littell, R.C. (1998). Projected
#'   multivariate linear models for directional data. Journal of the American
#'   Statistical Association, 93 (443), 1068 - 1077.
#'
#' @examples
#' library(bpnreg)
#' bpnr(Phaserad ~ Cond + AvAmp, Motor)
#'
#' @export
#'

bpnr <- function(pred.I, data, pred.II = pred.I,
                 its = 1000, burn = 1, n.lag = 1,
                 seed = NULL){

  if(sum(complete.cases(data))/length(complete.cases(data)) != 1){
    stop("The data contains missing values.")
  }

  if (!is.null(seed)){set.seed(seed)}

  mm <- mmr(pred.I, data, pred.II)

  output <- pnr(mm$theta, mm$XI, mm$XII, its, n.lag, burn)

  class(output) <- c("bpnr", class(output))

  summary.stats <- sumr(output, mm)

  output$beta1 <- summary.stats$beta1
  output$beta2 <- summary.stats$beta2
  output$a.x <- summary.stats$a.x
  output$a.c <- summary.stats$a.c
  output$b.c <- summary.stats$b.c
  output$SAM <- summary.stats$SAM
  output$AS <- summary.stats$AS
  output$SSDO <- summary.stats$SSDO
  output$circ.diff <- summary.stats$circ.diff

  output$Call <- match.call()
  output$lin.coef.I <- summary.stats$lin.res.I
  output$lin.coef.II <- summary.stats$lin.res.II
  output$circ.coef <- summary.stats$circ.res
  output$circ.coef.cat <- summary.stats$circ.res.cat
  output$circ.coef.means <- summary.stats$circ.res.means
  output$model.fit <- summary.stats$model.fit
  output$mm <- mm

  output

}

#' Fit a Bayesian circular mixed-effects model
#'
#' This function fits a Bayesian circular mixed-effects model based on the
#' projected normal distribution.
#'
#' @param pred.I model equation for effects of component 1.
#' @param data the dataframe used for analysis.
#' @param pred.II model equation for effects of component 2.
#' @param its output iterations of the MCMC sampler.
#' @param burn number of burn-in iterations.
#' @param n.lag amount of lag for the iterations and burn-in.
#' @param seed user-specified random seed.
#'
#' @details Because the model is based on the projected normal distribution, a
#'   model equation has to be given for the fixed and random effects of the two
#'   components. By default the model equation of the second component
#'   \code{pred.II} is set to be equal to that of the first component. Note that
#'   the circular outcome needs to be measured in radians on a scale from 0 to
#'   2\eqn{\pi}. For more information about the projected normal distribution
#'   see Presnell, Morrisson & Littell (1998). The model can handle at most one
#'   grouping factor.
#'
#'   A tutorial on how to use this function can be found in Cremers & Klugkist
#'   (2018). More details on the sampling algorithm and interpretation of the
#'   coefficients from the model can be found in Nuñez-Antonio & Guttiérrez-Peña
#'   (2014) and Cremers, Pennings, Mainhard & Klugkist (2019). The uninformative priors
#'   for the fixed effect regression coefficients of the two components are set to N(0, 10000).
#'   Note that the model is only developed for models with a single nesting variable.
#'
#' @return A \code{bpnme} object, which can be further analyzed using the
#'   associated functions \code{\link{traceplot.bpnme}},
#'   \code{\link{BFc.bpnme}}, \code{\link{coef_lin.bpnme}},
#'   \code{\link{coef_circ.bpnme}}, \code{\link{coef_ran.bpnme}},
#'   \code{\link{fit.bpnme}} and \code{\link{print.bpnme}}.
#'
#'   A \code{bpnr} object contains the following elements (some elements are not
#'   returned if not applicable)
#'
#'   \describe{ \item{\code{beta1}}{A matrix of posterior samples for the fixed
#'   effects coefficients for the first component.} \item{\code{beta2}}{A
#'   matrix of posterior samples for the fixed effects coefficients for the
#'   second component.} \item{\code{b1}}{An array of posterior samples for the
#'   random effects coefficients for the first component.} \item{\code{b2}}{An
#'   array of posterior samples for the random effects coefficients for the
#'   second component.} \item{\code{omega1}}{An array of posterior samples for
#'   the random effect variances of the first component.}
#'   \item{\code{omega2}}{An array of posterior samples for the random effect
#'   variances of the second component.} \item{\code{predictiva}}{A list
#'   containing the posterior density values for all timepoints of individuals
#'   in the dataset for all iterations. The rowsums of this matrix are the
#'   likelihood values for all iterations} \item{\code{circular.ri}}{A vector of
#'   posterior samples for the circular random intercepts.}
#'   \item{\code{N}}{Number of observed cases.} \item{\code{its}}{Number of
#'   output iterations.} \item{\code{n.lag}}{One in \code{n.lag} iterations will
#'   be saved as output iteration. Set lag to 1 to save all iterations
#'   (default).} \item{\code{burn}}{Burn-in time for the MCMC sampler.}
#'   \item{\code{p1}}{Number of fixed effect parameters predicting the first
#'   component.} \item{\code{p2}}{Number of fixed effect parameters predicting
#'   the second component.} \item{\code{q1}}{Number of random effect parameters
#'   predicting the first component.} \item{\code{q2}}{Number of random effect
#'   parameters predicting the second component.} \item{\code{a.x}}{A matrix of
#'   posterior samples for \code{a.x} which describes the location of the
#'   inflection point of the regression curve on the axis of the predictor.}
#'   \item{\code{a.c}}{A matrix of posterior samples for \code{a.c} which
#'   describes the location of the inflection point of the regression curve on
#'   the axis of the circular outcome.} \item{\code{b.c}}{A matrix of posterior
#'   samples for \code{b.c} which describes the slope of the tangent line at the
#'   inflection point.} \item{\code{SAM}}{A matrix of posterior samples for the
#'   circular regression slopes at the mean.} \item{\code{AS}}{A matrix of
#'   posterior samples for the average slopes of the circular regression.}
#'   \item{\code{SSDO}}{A matrix of posterior samples for the signed shortest
#'   distance to the origin.} \item{\code{circ.diff}}{A matrix of posterior
#'   samples for the circular difference found between levels of categorical
#'   variables and the intercept.} \item{\code{cRSnum}}{A vector of posterior
#'   samples of the circular random slope estimates for the continuous
#'   variables} \item{\code{cRScat}}{A vector of posterior samples of the
#'   circular random slope estimates for the categorical variables}
#'   \item{\code{cRS}}{A vector of posterior samples of the circular random
#'   slope estimates} \item{\code{cRI}}{A vector of posterior samples of the
#'   mean resultant length of the circular random intercept, a measure of
#'   concentration.}
#'   \item{\code{Call}}{The matched call.} \item{\code{lin.coef.I}}{The mean,
#'   mode, standard deviation and 95 % confidence interval of the highest
#'   posterior density of the linear fixed effect coefficients for \code{beta1}.}
#'   \item{\code{lin.coef.II}}{The mean, mode, standard deviation and 95 %
#'   confidence interval of the highest posterior density of the linear fixed
#'   effect coefficients for \code{beta2}.} \item{\code{circ.coef}}{The mean, mode,
#'   standard deviation and 95 % confidence interval of the highest posterior
#'   density for \code{a.x}, \code{a.c}, \code{SSDO}, and the circular fixed
#'   effect coefficients \code{b.c}, \code{AS}, and \code{SAM}}
#'   \item{\code{circ.coef.cat}}{The mean, mode, standard deviation and 95 %
#'   confidence interval of the highest posterior density the circular
#'   difference between levels of categorical variables and the intercept.}
#'   \item{\code{circ.coef.means}}{The mean, mode, standard deviation and 95 %
#'   confidence interval of the highest posterior density of circular means of
#'   the categorical variables.} \item{\code{model.fit}}{A list of information
#'   criteria for assessment of model fit.} \item{\code{lin.res.varrand.I}}{The
#'   mean, mode, standard deviation and 95 % confidence interval of  the
#'   variances of the random intercepts and slopes of component I.}
#'   \item{\code{lin.res.varrand.II}}{The mean, mode, standard deviation and 95
#'   % confidence interval of the variances of the random intercepts and slopes
#'   of component II.} \item{\code{circ.res.varrand}}{The mean, mode, standard
#'   deviation and 95 % confidence interval of the circular variances of the
#'   random intercepts and slopes.} \item{\code{mm}}{A list of information,
#'   model matrices, sample size, etc. on the specified model.} }
#'
#' @source Cremers, J., Mainhard, M.T. & Klugkist, I. (2018). Assessing a
#'   Bayesian Embedding Approach to Circular Regression Models. Methodology, 14, 69-81.
#'
#' @source Cremers, J. & Klugkist, I. (2018). One direction? A tutorial for
#'   circular data with examples in cognitive psychology. Frontiers in
#'   Psychology: Cognitive Science.
#'
#' @source Cremers, J., Pennings, H.J.M., Mainhard, M.T. & Klugkist, I. (2019).
#'   Circular Modelling of Circumplex Measurements for Interpersonal Behavior.
#'   Assessment, Online First.
#'
#' @source Nuñez-Antonio, G. & Gutiérrez-Peña, E. (2014). A Bayesian model for
#'   longitudinal circular data based on the projected normal distribution.
#'   Computational Statistics and Data Analysis, 71, 506-519.
#'
#' @source Presnell, B., Morrison, S.P. & Littell, R.C. (1998). Projected
#'   multivariate linear models for directional data. Journal of the American
#'   Statistical Association, 93 (443), 1068 - 1077.
#'
#' @examples
#' library(bpnreg)
#' bpnme(Error.rad ~ Maze + Trial.type + (1|Subject), Maps, its = 100)
#'
#' @export

bpnme <- function(pred.I, data, pred.II = pred.I,
                  its = 1000, burn = 1, n.lag = 1,
                  seed = NULL){

  if(sum(complete.cases(data))/length(complete.cases(data)) != 1){
    stop("The data contains missing values.")
  }

  if (!is.null(seed)){set.seed(seed)}

  mm <- mmme(pred.I, data, pred.II)

  circular.ri <- matrix(NA, mm$N, its)

  predictiva <- list()

  for (i in 1:mm$N){
    predictiva[[i]] <- matrix(NA, its, mm$no.Meas[i])
  }

  output <- pnme(lapply(mm$theta, cos), lapply(mm$theta, sin),
                 mm$XI, mm$XII, mm$ZI, mm$ZII, mm$ZtZI, mm$ZtZII,
                 mm$R, predictiva,
                 its = its, burn = burn, lag = n.lag, mm$N)

  class(output) <- c("bpnme", class(output))

  rownames(output$beta1) <- colnames(mm$XI[[1]])
  rownames(output$beta2) <- colnames(mm$XII[[1]])
  colnames(output$b1) <- colnames(mm$ZI[[1]])
  colnames(output$b2) <- colnames(mm$ZII[[1]])
  colnames(output$omega1) <- colnames(mm$ZI[[1]])
  rownames(output$omega1) <- colnames(mm$ZI[[1]])
  colnames(output$omega2) <- colnames(mm$ZII[[1]])
  rownames(output$omega2) <- colnames(mm$ZII[[1]])

  # Compute circular RI

  for (i in 1:mm$N){

    if(!("(Intercept)" %in% colnames(mm$XI[[i]])) & !("(Intercept)" %in% colnames(mm$XII[[i]]))){

      circular.ri[i,] <- atan2(0 + output$b2[i,1,], 0 + output$b1[i,1,])

    }else if(!("(Intercept)" %in% colnames(mm$XI[[i]]))){

      circular.ri[i,] <- atan2(output$beta2[1,] + output$b2[i,1,], 0 + output$b1[i,1,])

    }else if(!("(Intercept)" %in% colnames(mm$XII[[i]]))){

      circular.ri[i,] <- atan2(0 + output$b2[i,1,], output$beta1[1,] + output$b1[i,1,])

    }else{

      circular.ri[i,] <- atan2(output$beta2[1,] + output$b2[i,1,], output$beta1[1,] + output$b1[i,1,])

    }
  }

  output[["circular.ri"]] = circular.ri
  output$beta1 = t(output$beta1)
  output$beta2 = t(output$beta2)

  summary.stats <- summe(output, mm)

  output$a.x <- summary.stats$a.x
  output$a.c <- summary.stats$a.c
  output$b.c <- summary.stats$b.c
  output$SAM <- summary.stats$SAM
  output$AS <- summary.stats$AS
  output$SSDO <- summary.stats$SSDO
  output$circ.diff <- summary.stats$circ.diff
  output$cRSnum <- summary.stats$varrand.num
  output$cRScat <- summary.stats$varrand.cat
  output$cRS <- data.frame(summary.stats$varrand.cat, summary.stats$varrand.num)
  output$cRI <- summary.stats$circ.varrand.ri

  output$Call <- match.call()
  output$lin.coef.I <- summary.stats$lin.res.I
  output$lin.coef.II <- summary.stats$lin.res.II
  output$circ.coef <- summary.stats$circ.res
  output$circ.coef.cat <- summary.stats$circ.res.cat
  output$circ.coef.means <- summary.stats$circ.res.means
  output$model.fit <- summary.stats$model.fit
  output$lin.res.varrand.I <- summary.stats$lin.res.varrand.I
  output$lin.res.varrand.II <- summary.stats$lin.res.varrand.II
  output$circ.res.varrand <- summary.stats$circ.res.varrand
  output$mm <- mm

  output

}

